i#2387 assert in drcov: fixed a bug in drcov for module reload

zhaoqin opened the pull request:
This CL fixed a bug in drcov handling module reload.
Because there could be more than one submodules of a module, we should iterate
the whole module table to update all submodules when a module is reloaded.

Fixes #2387

zhaoqin commented:
Test to be added after i2006 PR commit, since the CL may conflict.

derekbruening commented on ext/drcovlib/modules.c:
```diff
@@ -174,11 +176,12 @@ event_module_load(void *drcontext, const module_data_t *data, bool loaded)
 #endif
-            break;
+            /* We may just see a sub module, so we cannot stop here. We need iterate
+             * all entries to update all sub modules of the reloaded module.
```
I don't understand: "We may just see a sub module": you mean if the app does a custom mmap of just one segment of an ELF file, and then goes and executes from that segment to trigger the module load event?  But why would that require further iteration?  All non-contiguous entries are guaranteed to be in the vector immediately following the initial entry.  Why would we need to iterate further?

zhaoqin replied:
The bug actually happens like below:
1. a module with multiple sub-modules is loaded, the module self and all sub-modules are added into module_table.
2. the module is unloaded, the module and all its sub-modules are marked as unloaded
3. the module is loaded at the same place (where the bug happens)
    the first entry we found match is actually a sub-module, and there are more sub-modules located before this submodule need update, but the code only updated all module/sub-modules after.
We could fixed the bug by forward scan in the outer loop, so we will see the module first before any submodule.
I will see if there is a better fix so we can reuse the code of handling the non-contiguous module reload.

derekbruening replied:
The first segment entry should match as all segments share the same module_data_t pointer.  Why didn't the first one match?

zhaoqin replied:
Because the loop is backward.

zhaoqin commented:
Updated commit message.

i#2387 assert in drcov: fixed a bug in drcov for module reload

This CL fixed a bug in drcov handling module reload.
Because a module could have multiple sub-modules, we should find the
containing module and update all its sub-modules on reload.

Fixes #2387

derekbruening commented on ext/drcovlib/modules.c:
```diff
@@ -141,6 +141,8 @@ event_module_load(void *drcontext, const module_data_t *data, bool loaded)
         entry = drvector_get_entry(&module_table.vector, i);
         mod   = entry->data;
         if (entry->unload &&
+            /* only check the main (containing) module */
```
Please add a note that this is needed b/c the loop is backward

zhaoqin replied:
done

derekbruening commented on ext/drcovlib/modules.c:
```diff
@@ -297,7 +299,8 @@ event_module_unload(void *drcontext, const module_data_t *data)
         entry = drvector_get_entry(&module_table.vector, i);
         ASSERT(entry != NULL, "fail to get module entry");
-        if (pc_is_in_module(entry, data->start))
+        /* only check the main (containing) module */
```
Please add a note that this is needed b/c the loop is backward

zhaoqin replied:
done

derekbruening commented on ext/drcovlib/modules.c:
```diff
@@ -234,7 +236,7 @@ static inline void
 {
     if (mod_index != NULL)
-        *mod_index = entry->id;
+        *mod_index = entry->containing_id; /* Yes, the main (containing) module. */
```
I don't think this is what we want: drmodtrack treats the segments as separate entries and exposes that to the user.  The user expects the resulting index to actual contain the queried PC, which is not true with this change.  What is your reason for wanting to change this?

zhaoqin replied:
removed.
I want to return the consistent result.
Currently, it could return either the segment or the containing module.

derekbruening replied:
In terms of the bounds they are all segments: no entry has the whole bounds, so in that sense it is consistent.  The "containing" module entry is simply the first segment.

derekbruening commented on ext/drcovlib/modules.c:
```diff
@@ -234,7 +238,7 @@ static inline void
 {
     if (mod_index != NULL)
-        *mod_index = entry->id;
+        *mod_index = entry->id; /* We expose the segment. */
```
It has to be exposed: some users will not work properly otherwise.  I think you were not aware of what is exposed: see drmodtrack_info_t's docs.

zhaoqin commented:
should I update the comment?

On Thu, Apr 27, 2017 at 1:57 PM, Derek Bruening <notifications@github.com>
wrote:

> *@derekbruening* approved this pull request.
> ------------------------------
>
> In ext/drcovlib/modules.c
> <https://github.com/DynamoRIO/dynamorio/pull/2391#discussion_r113762089>:
>
> > @@ -234,7 +238,7 @@ static inline void
>  lookup_helper_set_fields(module_entry_t *entry, OUT uint *mod_index, OUT app_pc *mod_base)
>  {
>      if (mod_index != NULL)
> -        *mod_index = entry->id;
> +        *mod_index = entry->id; /* We expose the segment. */
>
> It has to be exposed: some users will not work properly otherwise. I think
> you were not aware of what is exposed: see drmodtrack_info_t's docs.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2391#pullrequestreview-35179020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkcDOlLAIOu9_o4qqAuLDDep48KM_QAks5r0NbsgaJpZM4NKYlC>
> .
>

derekbruening commented:
Doesn't matter, just seems a little odd.

zhaoqin commented:
sorry, saw you message after merge

fhahn commented:
Looks like this is a stale branch, I'll delete it
